chronicle-queue/src/test/scala/akka/stream/alpakka/chroniclequeue/ChronicleQueueAtLeastOnceSpec.scala

```diff
@@ -0,0 +1,355 @@
+    clean()
+  }
+
+  it should "buffer for a throttled stream" in {
+    var t1, t2 = Long.MinValue
+    val util = new StreamSpecUtil[T, Event[T]]
+    import util._
+    val buffer = ChronicleQueueAtLeastOnce[T](config)
+    val t0 = System.nanoTime
+    def counter(recordFn: Long => Unit) =
+      Flow[Any]
+        .map(_ => 1L)
+        .reduce(_ + _)
+        .map { s =>
+          recordFn(System.nanoTime - t0)
+          s
+        }
+        .toMat(Sink.head)(Keep.right)
+
+    val streamGraph = RunnableGraph.fromGraph(GraphDSL.create(counter(t1 = _), flowCounter)((_, _)) {
+      implicit builder => (sink, total) =>
+        import GraphDSL.Implicits._
+        val bc1 = builder.add(Broadcast[T](2))
+        val bc2 = builder.add(Broadcast[Event[T]](2))
+        val commit = buffer.commit[T] // makes a dummy flow if autocommit is set to false
+        in ~> transform ~> bc1 ~> buffer.async ~> throttle ~> commit ~> bc2 ~> sink
+        bc2 ~> total
+        bc1 ~> counter(t2 = _)
+        ClosedShape
+    })
+    val (countF, totalF) = streamGraph.run()
+    val count = Await.result(countF, awaitMax)
+    val totalProcessed = Await.result(totalF, awaitMax)
+    eventually { buffer.queue shouldBe 'closed }
+
+    println("Time difference (ms): " + (t1 - t2) / 1000000d)
+    println(s"Total count $count vs total processed $totalProcessed")
+    count shouldBe elementCount
+    totalProcessed shouldBe elementCount
+    t1 should be > t2 // Give 6 seconds difference. In fact, it should be closer to 9 seconds.
+    clean()
+  }
+
+  it should "recover from unexpected stream shutdown" in {
+    implicit val util = new StreamSpecUtil[T, Event[T]]
+    import util._
+
+    val mat = ActorMaterializer()
+    val pBufferInCount = new AtomicInteger(0)
+    val commitCount = new AtomicInteger(0)
+    val finishedGenerating = Promise[Done]
+    val counter = new AtomicInteger(0)
+
+    def fireFinished() =
+      Flow[T]
+        .map { e =>
+          if (counter.incrementAndGet() == failTestAt) finishedGenerating.success(Done)
+          e
+        }
+        .toMat(Sink.ignore)(Keep.right)
+
+    val shutdownF = finishedGenerating.future.map { d =>
+      mat.shutdown(); d
+    }
+
+    val graph = RunnableGraph.fromGraph(GraphDSL.create(Sink.ignore) { implicit builder => sink =>
+      import GraphDSL.Implicits._
+      val buffer = ChronicleQueueAtLeastOnce[T](config)
+        .withOnPushCallback(() => pBufferInCount.incrementAndGet())
+        .withOnCommitCallback(() => commitCount.incrementAndGet())
+      val commit = buffer.commit[T] // makes a dummy flow if autocommit is set to false
+      val bc = builder.add(Broadcast[T](2))
+
+      in ~> transform ~> bc ~> buffer.async ~> throttle ~> commit ~> sink
+      bc ~> fireFinished()
+
+      ClosedShape
+    })
+    val sinkF = graph.run()(mat)
+    Await.result(shutdownF, awaitMax)
+    Await.result(sinkF.failed, awaitMax) shouldBe an[AbruptTerminationException]
```

Comment:
In our case, `BroadcastBufferSpec` and all its permutations were sporadically failing in Travis. It turns out the stream has not fully stopped by the time you get this exception. So we needed to wait for counters to be stable before we continue. This has not affected this test just yet but is just a warning. You may want to consider adding that wait, too. Please see https://github.com/paypal/squbs/blob/master/squbs-pattern/src/test/scala/org/squbs/pattern/stream/BroadcastBufferSpec.scala#L155

Reply:
I will keep this in mind thanks